File: AngularJS/src/org/angular2/settings/AngularSettings.java
package org.angular2.settings;

import com.intellij.lang.typescript.compiler.ui.TypeScriptServerServiceSettings;
import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.ModificationTracker;
import com.intellij.openapi.util.SimpleModificationTracker;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;


@State(name = "AngularJSSettings", storages = @Storage("angular-settings.xml"))
public class AngularSettings implements
                             TypeScriptServerServiceSettings,
                             PersistentStateComponent<AngularSettings.Settings> {

  private final SimpleModificationTracker myModificationTracker = new SimpleModificationTracker();  
  
  private Settings myState = new AngularSettings.Settings();
  

  public static AngularSettings get(Project project) {
    return ServiceManager.getService(project, AngularSettings.class);
  }


  static class Settings {
    public boolean useService = true;
  }

  @Nullable
  @Override
  public AngularSettings.Settings getState() {
    return myState;
  }

  @Override
  public void loadState(@NotNull AngularSettings.Settings state) {
    if (myState == null || myState.useService != state.useService) {
      myModificationTracker.incModificationCount();
    }
    myState = state;
  }

  @Override
  public boolean isUseService() {
    return myState.useService;
  }

  @Override
  public void setUseService(boolean useService) {
    if (myState.useService != useService) {
      myModificationTracker.incModificationCount();
    }
    myState.useService = useService;
  }

  @Override
  public String getTitle() {
    return "Angular Language Service";
  }

  @NotNull
  @Override
  public ModificationTracker getModificationTracker() {
    return myModificationTracker;
  }
}

File: Angular/angular-backend/src/org/angular2/options/AngularSettings.kt
// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.angular2.options

import com.intellij.javascript.util.JSLogOnceService
import com.intellij.lang.typescript.compiler.TypeScriptCompilerConfigUtil.isEffectiveUseTypesFromServer
import com.intellij.lang.typescript.compiler.TypeScriptCompilerSettings
import com.intellij.lang.typescript.compiler.ui.TypeScriptServiceRestartService
import com.intellij.lang.typescript.lsp.defaultPackageKey
import com.intellij.openapi.Disposable
import com.intellij.openapi.components.BaseState
import com.intellij.openapi.components.Service
import com.intellij.openapi.components.SimplePersistentStateComponent
import com.intellij.openapi.components.State
import com.intellij.openapi.components.Storage
import com.intellij.openapi.components.StoragePathMacros
import com.intellij.openapi.components.service
import com.intellij.openapi.diagnostic.logger
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Disposer
import com.intellij.openapi.util.registry.Registry
import org.jetbrains.annotations.TestOnly

fun getAngularSettings(project: Project): AngularSettings = project.service<AngularSettings>()

@TestOnly
fun configureAngularSettingsService(project: Project, disposable: Disposable, serviceSettings: AngularServiceSettings) {
  val angularSettings = getAngularSettings(project)
  val old = angularSettings.serviceType
  angularSettings.serviceType = serviceSettings

  Disposer.register(disposable) {
    angularSettings.serviceType = old
  }
}

@Service(Service.Level.PROJECT)
@State(name = "AngularSettings", storages = [Storage(StoragePathMacros.WORKSPACE_FILE)])
class AngularSettings(val project: Project) : SimplePersistentStateComponent<AngularSettingsState>(AngularSettingsState()) {

  companion object {
    private val LOG = logger<AngularSettings>()
  }

  var serviceType: AngularServiceSettings
    get() = state.innerServiceType
    set(value) {
      val prevServiceType = state.innerServiceType
      state.innerServiceType = value
      if (prevServiceType != value) {
        project.service<TypeScriptServiceRestartService>().restartServices(
          isEffectiveUseTypesFromServer(prevServiceType == AngularServiceSettings.AUTO, useTypesFromServer)
            != isEffectiveUseTypesFromServer(state.innerServiceType == AngularServiceSettings.AUTO, useTypesFromServer))
      }
    }

  val useTypesFromServer: Boolean
    get() {
      val result =
        TypeScriptCompilerSettings.useTypesFromServerInTests
        ?: useServicePoweredTypesManualOverride
        ?: (Registry.`is`("angular.service.powered.type.engine.enabled.by.default") && state.useTypesFromServer)
      with(project.service<JSLogOnceService>()) {
        LOG.infoOnce { "'Service-powered type engine' option of AngularSettings: $result" }
      }
      return result
    }

  var useServicePoweredTypesManualOverride: Boolean?
    get() = when {
      state.useServicePoweredTypesEnabledManually -> true
      state.useServicePoweredTypesDisabledManually -> false
      else -> null
    }
    set(value) {
      val prevUseTypesFromServer = useTypesFromServer
      state.useServicePoweredTypesEnabledManually = value == true
      state.useServicePoweredTypesDisabledManually = value == false
      if (prevUseTypesFromServer != value) {
        project.service<TypeScriptServiceRestartService>().restartServices(
          isEffectiveUseTypesFromServer(serviceType == AngularServiceSettings.AUTO, prevUseTypesFromServer)
            != isEffectiveUseTypesFromServer(serviceType == AngularServiceSettings.AUTO, useTypesFromServer))
      }
    }
}

class AngularSettingsState : BaseState() {
  var innerServiceType: AngularServiceSettings by enum(AngularServiceSettings.AUTO)
  var packageName: String? by string(defaultPackageKey)
  @Deprecated(message = "Use useServicePoweredTypesEnabledManually and useServicePoweredTypesDisabledManually")
  var useTypesFromServer: Boolean by property(true)
  var useServicePoweredTypesEnabledManually: Boolean by property(false)
  var useServicePoweredTypesDisabledManually: Boolean by property(false)
}

enum class AngularServiceSettings {
  AUTO,
  DISABLED
}